prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -588,6 +592,28 @@ def build_transmission(method="sub2line", kwargs={"rounding": 3}):
     substations = filter_substations_with_zero_lines(hifld_substations)
     check_for_location_conflicts(substations)
 
+    # Filter out keyword arguments for get_mst_edges function
+    get_mst_kwargs = dict()
+    for keyword in {
+        "island_size_lower_bound",
+        "island_size_upper_bound",
+        "state_neighbor",
+        "min_dist_method",
+        "cost_metric",
+        "memory_efficient",
+        "kdtree_kwargs",
+    }:
+        if keyword in kwargs:
+            get_mst_kwargs[keyword] = kwargs.pop(keyword)
+        if "island_size_lower_bound" in get_mst_kwargs:
+            island_size_lower_bound = get_mst_kwargs["island_size_lower_bound"]
+        else:
+            island_size_lower_bound = 0
+        if "island_size_upper_bound" in get_mst_kwargs:
+            island_size_upper_bound = get_mst_kwargs["island_size_upper_bound"]
+        else:
+            island_size_upper_bound = len(substations) + 1
```

Comment:
I don't really understand the point of this section. Isn't the point of the current `build_transmission` `kwargs` parameter that it can take any combination of keyword arguments for different lower-level functions, and each function can be passed these combined keyword arguments and only care about the ones that are relevant to it, ignoring the rest?

Reply:
Yeah, I agree with the logic. However, we still need to filter out the corresponding kwargs for specific lower level functions, right? Otherwise, it will throw errors like `TypeError: connect_islands_with_minimum_cost() got an unexpected keyword argument 'rounding'.`

Reply:
I think we want to have the function definitions to have `**kwargs` all the way down to prevent this, i.e. `def connect_islands_with_minimum_cost(lines, subs, **kwargs)`.

Reply:
I think this logic looks good for passing `kwargs` through, however, when some `kwargs` hit the lowest level function, we should have them explicitly defined.

Reply:
I don't think that follows the current pattern in **transmission.py** though, since `map_lines_to_substations_using_coords` is a lowest-level function and still has `**kwargs`.

Reply:
I'm saying it follows the pattern in other packages such as `pandas`. 

Then we should refactor `map_lines_to_substations_using_coords` some day in the future once we have a clear idea on what `kwargs` it takes.

Reply:
Maybe I'm unclear on the design for the `build_transmission` then. Are we going to take `**kwargs` at the highest-level, and then separate this one dict into several lower-level dicts? If so, I think we could do this purely on keys, and then have a pattern of `dict.get(key, default value)` in the lower- (lowest-?) level functions.

Reply:
That's another option, then how do we write doc strings? I think there should be a classic way of handling `kwargs` and we can simply follow that. The current implementation is my 2 cents after looking through pandas documentation a little bit. 

Maybe @jon-hagg has a better idea on this?
The issue we encounter here is, given a top level user facing function, which takes a set of `kwargs`, however, different subsets of `kwargs` should be passed to different middle/lower level functions. What would be the classic pattern of function definitions for top level function, middle level functions, and lower level functions to handle this situdation?

Reply:
Btw, the two `if`s setting defaults for island size bounds here is on a different track, given they are also used in `transmission.py` as well.

Reply:
You raise a good point about the docstrings.

I think one approach would be to define a dictionary `{str: set of str}` within build_transmission that dictates which of the kwargs go to which lower-level functions, then decompose the highest-level `**kwargs` to lower-level ones (making sure that there are no leftover keys) and pass those as appropriate... but it gets confusing when some of those highest-level kwargs are going to be used in the highest-level function as well. What do you think about moving everything from `mst_edges = get_mst_edges(...)` to `lines = pd.concat([lines, new_lines])` to a lower-level function? E.g. `lines, substations = filter_connected_components_and_connect(lines, substations, **specific_kwargs)`?

Reply:
That's another work around which is cleaner. However, given it is a general implementation related question, I presume there should be a classic way to handle this. Let's put it on hold waiting for ideas from people with related experiences for now.

Reply:
Regarding the refactor idea, I like it. The down side is the MST feature is tied up with this framework, which makes it less flexible to use it elsewhere and it also closes the window of further processing on the selected edges in the main function as well.

Reply:
Haven't fully thought this through, but one way to handle the various options is using classes, and constructing the top level instance based on the mst algorithm. Part of this could include also using a class to define/validate the kwargs that would apply to the algorithm. Something like

```
class Transmission:
    mst_builder: MstBuilder

    def __init__(self, mst_algo="naive", mst_options=None)
        # set defaults or validate the combination of options/algo

    def build_transmission(self, method):
        # use the mst_builder instance

class NaiveMst:
    options: NaiveMstOptions

    def build(self):
        # do stuff

class KDTreeMst:
    options: KDTreeMstOptions

    def build(self):
        # do stuff

class KDTreeMstOptions:
    def __init__(self, island_size_lower_bound, island_size_upper_bound, other_params):
        # set defaults or validate param types/values
        
    def build(self):
        # do stuff

builder = Transmission(KDTreeMst(KDTreeMstOptions()))


```

Reply:
> Haven't fully thought this through, but one way to handle the various options is using classes, and constructing the top level instance based on the mst algorithm. Part of this could include also using a class to define/validate the kwargs that would apply to the algorithm. Something like
> 
> ```
> class Transmission:
>     mst_builder: MstBuilder
> 
>     def __init__(self, mst_algo="naive", mst_options=None)
>         # set defaults or validate the combination of options/algo
> 
>     def build_transmission(self, method):
>         # use the mst_builder instance
> 
> class NaiveMst:
>     options: NaiveMstOptions
> 
>     def build(self):
>         # do stuff
> 
> class KDTreeMst:
>     options: KDTreeMstOptions
> 
>     def build(self):
>         # do stuff
> 
> class KDTreeMstOptions:
>     def __init__(self, island_size_lower_bound, island_size_upper_bound, other_params):
>         # set defaults or validate param types/values
>         
>     def build(self):
>         # do stuff
> 
> builder = Transmission(KDTreeMst(KDTreeMstOptions()))
> ```

Not sure I understand what we are trying to solve by changing everything into class structure. In this situation, we don't have specific data structures with different set of functions associated with it. It fits well with a sequence of functions calls to process the data in a proposed way. Also, I don't see how class instance might help passing `kwargs` through a nested function call structure assuming we still have one user facing function.

Reply:
Whether or not classes are the way to go here, I would say in general they _may_ be useful regardless of if there is a conventional data structure to implement. E.g. having an "options" class could store a group of related kwargs and validate them upon creation, keeping that logic separate from the rest of the function (which makes it testable independently, if applicable), and clearly shows what the grouping is.

In this case, my intuition is to use classes to represent the polymorphic aspect (naive vs kdtree, or various sets of kwargs). Constructing the object up front allows internal calls to operate within a larger scope, without having to have a bunch of pass-through parameters. It essentially moves some if/else logic to the top layer, rather than having it embedded in lower level functions. So not a fundamental reduction in complexity, but might improve readability and ease of development. 

Feel free to use any/all/none of the class approach. If we still want the user interface level to be function based, it could still use classes internally, or not, and construct the appropriate instances based on the function arguments.

Reply:
If we keep the user interface level to be function based, but go for classes internally, we still need the user facing function to pass through the `kwargs` upon internal class instance creation, right?

Reply:
I think so, yes. Not sure if it's applicable here, but for example if we had like 10 layers of nested function calls, creating a class instance (specifically, the object graph) up front would save on a bunch of redundant parameters, and only require parsing kwargs into objects once. Since we have like 2-3 layers, the trade off is not so obvious.